rfcs/013-node-types.md

```diff
@@ -0,0 +1,145 @@
+---
+name: Node Types
+about: Introduce heterogeneous node types to CouchDB 4
+title: 'Node Types'
+labels: rfc, discussion
+assignees: ''
+
+---
+
+# Introduction
+
+This RFC proposes the ability to have different node types in CouchDB 4+. This
+would improve performance and allow for a more efficient use of resources.
+
+## Abstract
+
+Previously, in CouchDB 2 and 3, cluster functionality was uniformly distributed
+amongst the nodes. Any node could accept HTTP requests, run replication jobs
+and build secondary indices. With the FDB-based topology, CRUD operations have
+lower resource needs and so it could be useful to have a heterogeneous
+topology, where for example, CRUD operations run on lower capacity nodes, and a
+few higher capacity nodes handle replication or indexing jobs.
+
+
+## Requirements Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+*node type* : A label used to designate a subset of CouchDB functionality.
+
+---
+
+# Detailed Description
+
+## Node Types
+
+A node type is a string description of a some internal CouchDB
+functionality. These are the initially defined node types:
+
+ * "api_frontend" : Indicates this node can accept HTTP API requests.
+ * "view_indexing" : Indicates this node can build map/reduce view indices.
+ * "search_indexing": Indicates this node can build search indices.
+ * "replication" : Indicates this node can run replication jobs.
```

Comment:
while we lose traditional compaction, IIRC that there might be a similar cleanup task involved with some of the proposed FDB data modelling. Would a “garbage_collection” role make sense here, too? I might be missing an evolution of the FDB data model that doesn’t produce garbage, in that case, ignore me :)

Reply:
Hmm. That does sound familiar but I don't recall the exact thing. One possibility there would to be a cleaner-upper to remove background jobs when a database which created them was removed. It could be done asynchronously, but since we haven't discussed that and we're not sure how it would work, I'd be inclined to skip it. Another label would trivial to add to the list though.

Reply:
I recall one option for ensuring that getting the latest rev of a doc is always a one-shot leaves garbage behind, I think it was proposed by @kocolosk, but that might easily have been optimised away in the design already. Making adding more roles simple is a good plus here. Thanks!